File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/BreadthFirstSearch/103_BinaryTreeZigzagLevelOrderTraversal.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-24 11:19:42
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
c.. Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        __(root__NULL){
            r_ {};
        }
        vector<vector<int>> ans;
        deque<TreeNode*> levels{root};
        bool order=true;
        _____(!levels.empty()){
            vector<int> cur_level;
            int l_size = levels.size();
            ___(int i=0;i<l_size;i++){
                TreeNode* cur_node = levels.front();
                levels.pop_front();
                cur_level.push_back(cur_node->val);
                __(cur_node->left)  levels.push_back(cur_node->left);
                __(cur_node->right) levels.push_back(cur_node->right);
            }
            __(!order)  reverse(cur_level.begin(),cur_level.end());
            ans.push_back(cur_level);
            order = !order;
        }
        r_ ans;
    }
};

/*
[]
[1]
[1,2,3]
[0,1,2,3,4,5,6,null,null,7,null,8,9,null,10]
*/

File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/BreadthFirstSearch/103_BinaryTreeZigzagLevelOrderTraversal.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

c.. Solution o..
    ___ zigzagLevelOrder  root
        __ n.. root:
            r_ []

        left2right = 1
        # 1. scan the level from left to right. -1 reverse.
        ans, stack, temp   # list, [root], []
        _____ stack:
            temp = [node.val ___ node __ stack]
            stack = [child ___ node __ stack
                     ___ child __ (node.left, node.right) __ child]

            ans += [temp[::left2right]]         # Pythonic way
            left2right *= -1

        r_ ans

"""
[]
[1]
[1,2,3]
[0,1,2,3,4,5,6,null,null,7,null,8,9,null,10]
"""
